leakpro/synthetic_data_attacks/syn_text_pii_scanner/data_handling.py

```diff
@@ -0,0 +1,265 @@
+            else:
+                aligned_labels = None
+            #Append EncodedDoc to processed_data
+            processed_data.append(EncodedDoc(
+                input_ids = encoding["input_ids"],
+                labels = aligned_labels,
+                offsets = encoding["offset_mapping"]
+            ))
+            #Check for max_length
+            max_length = max(len(encoding["input_ids"]), max_length)
+        return processed_data, max_length
+
+    def align_labels(self, *,
+        encoding: BatchEncoding,
+        annotations: List[Dict[str, Any]]
+    ) -> Tuple[List[int], List[str], List[Dict[str, Tuple[int, int]]]]:
+        """Align labels with tokenized input.
+
+        Args:
+            encoding (BatchEncoding): The tokenized input.
+            annotations (List[Dict[str, Any]]): The annotations for the input.
+
+        Returns:
+            List[int]: List of aligned labels
+
+        """
+        #Set variables
+        offsets_map = np.array(encoding["offset_mapping"])
+        len_seq = len(encoding["input_ids"])
+        aligned_labels: List[int] = [0] * len_seq  # 0 corresponds to "O" label
+        prefix_length = len(self.tokenizer.bos_token)
+        #Iterate through annotations
+        for anno in annotations:
+            #Annotated label key value is in label_set.labels
+            if anno[self.label_key] in self.label_set.labels:
+                #Calculate start and end token indexes
+                start = prefix_length + anno["start_offset"]
+                end = prefix_length + anno["end_offset"]
+                cond_start = offsets_map[:, 0] >= start
+                cond_end = offsets_map[:, 1] <= end
+                cond = cond_start & cond_end
+                matches = np.nonzero(cond)[0]
+                start_tok_idx = matches.min()
+                end_tok_idx = matches.max()
+                #Iterate through token indexes and set aligned labels depending on IOB2_FORMAT
+                for num, token_ix in enumerate(range(start_tok_idx, end_tok_idx+1)):
+                    if self.label_set.IOB2_FORMAT:
+                        prefix = "B" if num == 0 else "I"
+                        aligned_labels[token_ix] = self.label_set.labels_to_id[f"{prefix}-{anno[self.label_key]}"]
+                    else:
+                        aligned_labels[token_ix] = self.label_set.labels_to_id[anno[self.label_key]]
+        return aligned_labels
+
+    def __len__(self) -> int:
+        """Get the number of encoded documents.
+
+        Returns:
+            int: The number of encoded documents.
+
+        """
+        return len(self.data)
+
+    def __getitem__(self, idx: int) -> EncodedDoc:
+        """Get a single encoded document by index.
+
+        Args:
+            idx (int): The index of the encoded document.
+
+        Returns:
+            EncodedDoc: The encoded document at the given index.
+
+        """
+        return self.data[idx]
+
+@dataclass
+class CollatorWPadding:
```

Comment:
Didnt know it was possible to use dataclass in collator functions for the data loader. This makes it so much more flexible! Nice.